### PR TITLE
add new ``invalid-bases-assignment`` check

### LINKED ISSUE #584: [easy] New rules for assignment to __bases__
Originally reported by: **Claudiu Popa (BitBucket: [PCManticore](http://bitbucket.org/PCManticore), GitHub: @PCManticore)**

---
- can assign only tuples to `__bases__`

```
#!python

class A: pass
A.__bases__ = "lala"
```
- can assign only non empty tuples to `__bases__`

```
#!python

class A: pass
A.__bases__ = ()
```
- `__bases__` must be tuple of classes

```
#!python

class A: pass
A.__bases__ = (1, )
```

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/584

### PR DESCRIPTION
**Pull request by clavedeluna**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Pylint should warn a user if a class's `__bases__` dunder method is being assigned to anything other than a populated tuple. Python interpreter would raise a `TyperError` in these cases, for example

```
>>> class Apple:
...     pass
...
>>> Apple.__bases__ = "green"  
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can only assign tuple to Apple.__bases__, not str

>>> Apple.__bases__ = ()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can only assign non-empty tuple to Apple.__bases__, not ()

```

So this check would warn a pylint user about this before the interpreter would.

Closes #584

### CODE REVIEW

File: pylint/checkers/typecheck.py
Comment (clavedeluna) [inline pylint/checkers/typecheck.py:1322]: I also assumed HIGH is the right confidence here, lmk if that's not right.

**Comment by clavedeluna** [comment]
Maintainers:  are functional tests sufficient for new checkers, or should we also always add unit tests?

File: doc/data/messages/i/invalid-bases-assignment/good.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/i/invalid-bases-assignment/good.py:4]:
```suggestion
Apple.__bases__ = ("green", "red", )
```

**Review by DanielNoord — CHANGES_REQUESTED** [review]

File: tests/functional/i/invalid/invalid_bases_assignment.py
Comment (DanielNoord) [inline tests/functional/i/invalid/invalid_bases_assignment.py:16]: Why is this invalid?

File: tests/functional/i/invalid/invalid_bases_assignment.py
Comment (mbyrnepr2) [inline tests/functional/i/invalid/invalid_bases_assignment.py:12]: This one is a `TypeError`; perhaps we should make the [current definition](https://github.com/PyCQA/pylint/pull/7615/files#diff-9b76776694be02a09ca25a56adee0b3087ad5b86dfafd318923c4bf008c131c3R404) stricter.

File: tests/functional/i/invalid/invalid_bases_assignment.py
Comment (clavedeluna) [reply tests/functional/i/invalid/invalid_bases_assignment.py:16]: HA as I wrote this I inverted what's allowed for this check in my mind! Of course this should've passed. Will take a look.

File: tests/functional/i/invalid/invalid_bases_assignment.py
Comment (mbyrnepr2) [reply tests/functional/i/invalid/invalid_bases_assignment.py:12]: [Python docs](https://docs.python.org/3/c-api/object.html?highlight=__bases__) have it as _a tuple of base classes_

Comment (clavedeluna) [reply tests/functional/i/invalid/invalid_bases_assignment.py:12]: good catch!

File: tests/functional/i/invalid/invalid_bases_assignment.py
Comment (clavedeluna) [reply tests/functional/i/invalid/invalid_bases_assignment.py:16]: ah so turns out this is correct, both for the assigning to `x` and straight without assignment. it's because `Apple.__bases__ = ("red", "green")` gets a `TypeError: Apple.__bases__ must be tuple of classes, not 'str'` in the REPL. Also mentioned in a comment above  for ints. I'll update the definition for this checker to be more strict.

**Comment by clavedeluna** [comment]
This is trickier than we thought!

```
>>> class Fruit:
...     pass
...
>>> class Apple:
...     pass
...
>>> Apple.__bases__ = (Fruit, )
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __bases__ assignment: 'Fruit' deallocator differs from 'object'
```

Reading about, including https://stackoverflow.com/questions/3193158/bases-doesnt-work-whats-next

File: tests/functional/i/invalid/invalid_bases_assignment.py
Comment (DanielNoord) [reply tests/functional/i/invalid/invalid_bases_assignment.py:16]: Oke, but:
```python
bases = (int, str)
Apple__bases__ = bases
```

That works right?

Not to discourage you, but this is shaping up to be a lot more difficult than we might have originally envisioned.

Comment (clavedeluna) [reply tests/functional/i/invalid/invalid_bases_assignment.py:16]: Huh yeah that does work, didn't expect it. Let me summarize the findings and see how we can limit the scope here, if at all.

**Comment by clavedeluna** [comment]
@DanielNoord @Pierre-Sassoulas , maintainers et all:

Turns out this check is a bit more complicated than just "check if __bases__ is assigned to anything but a tuple". There are cases which also fail the interpreter that have to do way more with inheritance which I'm not sure is going to be easy to detect.

This is an example that will fail the interpreter
```
>>> class Tree:
...     pass
...
>>> class Shrub:
...     pass
...
>>> Shrub.__bases__
(<class 'object'>,)
>>> Shrub.__bases__ = (Tree,)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __bases__ assignment: 'Tree' deallocator differs from 'object'
```

However, these example will pass:

```
class Tree:
    pass

class Apple(Tree):
    pass

class Fruit(Tree):
    pass

Apple.__bases__ = (Fruit,)
Apple.__bases__ = (Tree, )
```

So in the Shrub first case, while we're assigning to a `tuple of classes` as the python doc states, these is an object allocation and inheritance issue.

To limit scope for this first iteration of this check, is it ok to implement the "naive" check of as long as __bases__ is being assigned to a tuple of classes, then good? This means pylint would for now miss the Shrub example.

**Comment by DanielNoord** [comment]
I'm not sure a "naive" check works here. I think it's very likely to create false positives here. The primer won't be able to give us much security as I don't think this is a very common pattern, which also doesn't help.

Personally I would prefer not to introduce such a difficult check if we haven't covered everything. We should probably investigate and come up with a good set of tests and work with those. `astroid` should have everything we need to check `mro's` etc.

**Comment by clavedeluna** [comment]
> I'm not sure a "naive" check works here. I think it's very likely to create false positives here. The primer won't be able to give us much security as I don't think this is a very common pattern, which also doesn't help.
> 
> Personally I would prefer not to introduce such a difficult check if we haven't covered everything. We should probably investigate and come up with a good set of tests and work with those. `astroid` should have everything we need to check `mro's` etc.

I agree with the sentiment that maybe adding a naive checker may not make sense, but only if we have examples of false positives. Given this is a very rare (hopefully!) pattern, I can't think of any false positives. If anything, like I mentioned above with the MRO issue, we would have _false negatives_.

**Comment by Pierre-Sassoulas** [comment]
Never used this pattern myself. How well does mypy or pyright handle this ? The easiest move might be not to play if they already do it well. I'm more and more wary of adding typing checks when other wildly adopted tools already do the jobs generically and easily.

**Comment by clavedeluna** [comment]
Alright, if you run this code

```
# pylint: disable=missing-module-docstring, missing-class-docstring
# pylint: disable=too-few-public-methods


class Tree:
    pass

class Shrub:
    pass

class Apple(Tree):
    pass

class Fruit(Tree):
    pass

Apple.__bases__ = (Fruit,)
Apple.__bases__ = (Tree, )

Tree.__bases__ = (Apple, )  # [invalid-bases-assignment]

Apple.__bases__ = ()  # [invalid-bases-assignment]
Apple.__bases__ = "lala"  # [invalid-bases-assignment]
Apple.__bases__ = True  # [invalid-bases-assignment]
Apple.__bases__ = (1,)  # [invalid-bases-assignment]
Apple.__bases__ = ("red", "green")  # [invalid-bases-assignment]

bases = ("orange", "yellow")
Apple.__bases__ = bases  # [invalid-bases-assignment]

Apple.__bases__ = Apple  # [invalid-bases-assignment]
Apple.__bases__ = (Apple, )  # [invalid-bases-assignment]


class EverGreen:
    pass

bases = (int, str)
EverGreen__bases__ = bases
```

against `mypy`, the results are interesting

```
23: error: Incompatible types in assignment (expression has type "str", variable has type "Tuple[type, ...]")  [assignment]
24: error: Incompatible types in assignment (expression has type "bool", variable has type "Tuple[type, ...]")  [assignment]
25: error: Incompatible types in assignment (expression has type "Tuple[int]", variable has type "Tuple[type, ...]")  [assignment]
26: error: Incompatible types in assignment (expression has type "Tuple[str, str]", variable has type "Tuple[type, ...]")  [assignment]
29: error: Incompatible types in assignment (expression has type "Tuple[str, str]", variable has type "Tuple[type, ...]")  [assignment]
31: error: Incompatible types in assignment (expression has type "Type[Apple]", variable has type "Tuple[type, ...]")  [assignment]
38: error: Incompatible types in assignment (expression has type "Tuple[Type[int], Type[str]]", variable has type "Tuple[str, str]")  [assignment]
```

this is telling me that mypy is actually doing a "naive" check as well. If it were a smart check against MRO, it should've detected
`Tree.__bases__ = (Apple, )  # [invalid-bases-assignment]`
or even easier
`Apple.__bases__ = (Apple, )  # [invalid-bases-assignment]`

because while this is assigning to a tuple of classes, it breaks inheritance.

@Pierre-Sassoulas @DanielNoord as maintainers could you decide what to do here?
1. do a naive approach as appears to be done in mypy (only check against tuple of classes, nothing else regarding inheritance)
2. close this PR, update issue (close it?) explaining why we won't do it

**Comment by Pierre-Sassoulas** [comment]
I'm mixed, as mypy is also doing it naively, it might be valuable to do it in pylint (maybe better later?). Maybe mypy is doing it really easily and we're going to have a hard time maintaining our version later. On the other a hand it's a shame to throw a check from an issue that has been opened a long time. In any case we should salvage the unrelated confidence change and refactor. Let's gather other feedback from maintainers.

**Comment by DanielNoord** [comment]
If we can't figure out how to do the non-naive check we don't provide much else compared to `mypy`. So for me I'd say either we implement the full check or leave this to `mypy`.

**Pull request closed without merging**
